### PR TITLE
use privileged containers for builds and docker registry

### PR DESCRIPTION
**Pull request by bparees**

### CODE REVIEW

**Comment by bparees** [comment]
Attempt to resolve https://github.com/openshift/origin/issues/263 for now.

@smarterclayton it looks like there's an upstream change coming that would allow setting of privileged container permissions via CLI, but we don't have it here yet...  forcing it to true is ugly, but if our builds aren't going to work without it, it doesn't seem that terrible.

if you're ok with this, i'll add changes to the doc to this PR to remove the selinux requirement.

Note that i had to make docker-registry privileged too, to deal w/ it's inability to write to /tmp/openshift.local.registry.

**Comment by bparees** [comment]
@smarterclayton readmes updated, please review

**Comment by bparees** [comment]
[test]

File: README.md [hunk 1]
```diff
@@ -14,6 +14,17 @@ NOTE: OpenShift is in alpha and is not intended for production use yet. However
 [![GoDoc](https://godoc.org/github.com/openshift/origin?status.png)](https://godoc.org/github.com/openshift/origin)
 [![Travis](https://travis-ci.org/openshift/origin.svg?branch=master)](https://travis-ci.org/openshift/origin)
 
+Security Warning
+----------------
+OpenShift no longer requires selinux to be disabled, however OpenShift is a system which runs docker containers on your system.  In some cases (build operations and the registry service) it does so using privileged containers.  Furthermore those containers access your host's docker daemon and perform `docker build` and `docker push` operations.  As such, you should be aware of the inherent security risks associated with performing `docker run` operations on arbitrary images as they effectively have root access.  This is particularly relevant when running the OpenShift nodes directly on your host system.
```
Comment (smarterclayton) [inline README.md]: SELinux

Comment (smarterclayton) [inline README.md]: Docker

File: README.md [hunk 2]
```diff
@@ -14,6 +14,17 @@ NOTE: OpenShift is in alpha and is not intended for production use yet. However
...
+
+For more information, see these articles:
+
+* http://opensource.com/business/14/7/docker-security-selinux
+* https://docs.docker.com/articles/security/
+
+The OpenShift security model will continue to evolve and tighten as we head towards production ready code.
```
Comment (smarterclayton) [inline README.md]: Very succint.

File: pkg/build/controller/strategy/custom.go [hunk 3]
```diff
@@ -48,9 +48,10 @@ func (bs *CustomBuildStrategy) CreateBuildPod(build *buildapi.Build) (*kapi.Pod,
 		Spec: kapi.PodSpec{
 			Containers: []kapi.Container{
 				{
-					Name:  "custom-build",
-					Image: strategy.Image,
-					Env:   containerEnv,
+					Name:       "custom-build",
+					Image:      strategy.Image,
+					Env:        containerEnv,
+					Privileged: true,
```
Comment (smarterclayton) [inline pkg/build/controller/strategy/custom.go]: Add a TODO to all of these, with a link to #662

**Comment by bparees** [comment]
@smarterclayton comments addressed.

**Comment by bparees** [comment]
fixed validation failure.

**Comment by smarterclayton** [comment]
[merge]

File: examples/sample-app/docker-registry-config.json [hunk 4]
```diff
@@ -65,7 +65,8 @@
                         "readOnly":false
                       }
                     ],
-                    "command": ["sh", "-c", "REGISTRY_URL=${DOCKER_REGISTRY_SERVICE_HOST}:${DOCKER_REGISTRY_SERVICE_PORT} OPENSHIFT_URL=http://${KUBERNETES_SERVICE_HOST}:443/osapi/v1beta1 exec docker-registry"]
+                    "command": ["sh", "-c", "REGISTRY_URL=${DOCKER_REGISTRY_SERVICE_HOST}:${DOCKER_REGISTRY_SERVICE_PORT} OPENSHIFT_URL=http://${KUBERNETES_SERVICE_HOST}:443/osapi/v1beta1 exec docker-registry"],
+		    privileged: true
```
Comment (soltysh) [inline examples/sample-app/docker-registry-config.json]: @bparees shouldn't privileged be quoted?

**Comment by rhatdan** [comment]
```Note that i had to make docker-registry privileged too, to deal w/ it's inability to write to /tmp/openshift.local.registry. ```
Should not require a privileged container, just label this directory correctly.  BTW why is this in /tmp?

**Comment by rhatdan** [comment]
We should discuss the docker build problem.  Isn't this using two different docker commands? The actual process doing the docker build is not privileged, just the command that start docker build?
If you control the syntax of the command then docker build will run with confinement.  I have a bug to run docker build with SELinux confinement in addition to other confinements.

**Comment by smarterclayton** [comment]
It's a throwaway dir for reuse.  I spawned an issue to track a better solution, which was linked above.

File: examples/sample-app/docker-registry-config.json (hunk 4 above)
Comment (bparees) [inline examples/sample-app/docker-registry-config.json]: probably should have been, but apparently it's not required :)

**Comment by bparees** [comment]
@rhatdan this PR should have added privileged to the registry, it was missing quotes but for some reason that didn't produce any error in our tests.  Anyway i've fixed the missing quotes here: https://github.com/openshift/origin/pull/676

as for what needs to be privileged:
we need privileged access (for the docker socket) for a couple steps...  and it depends on the type of build, but basically pull, build, and push are all commands liable to be run by the builder container which therefore need the docker socket.

**Comment by rhatdan** [comment]
@bparees You could try docker run --securityopt label:disabled rather then --privileged, although this would not provide you much increase in security.  It would disable just the SELinux labeling of the container, but keep the Capability drop.

**Comment by bparees** [comment]
@rhatdan k8s doesn't offer a way to control securityopt that i know of, so the only simple change i can make is to mark the container privileged.

**Comment by rhatdan** [comment]
Ok, I guess we need to grow some knowlege of this into K8s.

**Comment by smarterclayton** [comment]
Yeah, that is part of the security context work.

----- Original Message -----

**Pull request merged**